pkg/kubelet/cm/deviceplugin/pod_devices.go

```diff
@@ -17,207 +17,523 @@ limitations under the License.
+		ret[p] = make(map[string]resourceMap)
+		for c, dmap := range cmap {
+			ret[p][c] = make(map[string]deviceInfo)
+			for r, dinfo := range dmap {
+				ret[p][c][r] = deviceInfo{
+					status:    dinfo.status,
+					ids:       dinfo.ids,
+					allocResp: dinfo.allocResp,
+				}
+			}
+		}
 	}
 	return ret
 }
 
-func (pdev podDevices) insert(podUID, contName, resource string, devices sets.String, resp *pluginapi.AllocateResponse) {
-	if _, podExists := pdev[podUID]; !podExists {
-		pdev[podUID] = make(containerDevices)
+// pods returns the set of Pods currently known.
+func (a *devAllocationImpl) pods() (r sets.String) {
+	a.RLock()
+	defer a.RUnlock()
+
+	r = sets.NewString()
+	for k := range a.registry {
+		r.Insert(k)
 	}
-	if _, contExists := pdev[podUID][contName]; !contExists {
-		pdev[podUID][contName] = make(resourceAllocateInfo)
+	return r
+}
+
+// hasPod checks if there are records associated with the given pod.
+func (a *devAllocationImpl) hasPod(pod string) bool {
+	a.RLock()
+	defer a.RUnlock()
+	_, found := a.registry[pod]
+	return found
+}
+
+// devices returns all devices already allocated or reserved, keyed by resource name
+func (a *devAllocationImpl) devices() map[string]sets.String {
+	a.RLock()
+	defer a.RUnlock()
+	return a.getDevices()
+}
+
+// getDevices is an internal function returning all devices allocated or
+// reserved
+func (a *devAllocationImpl) getDevices() map[string]sets.String {
+	ret := make(map[string]sets.String)
+	for _, cmap := range a.registry {
+		for _, rmap := range cmap {
+			for r, dinfo := range rmap {
+				if _, ok := ret[r]; !ok {
+					ret[r] = sets.NewString()
+				}
+				ret[r] = ret[r].Union(dinfo.ids)
+			}
+		}
 	}
-	pdev[podUID][contName][resource] = deviceAllocateInfo{
-		deviceIds: devices,
-		allocResp: resp,
+	return ret
+}
+
+// update updates the pod allocation registry by list of active pods.
+// The function returns a map from resource name to set of device IDs.
+func (a *devAllocationImpl) purge(activePods sets.String) map[string]sets.String {
+	a.Lock()
+	defer a.Unlock()
+
+	ret := make(map[string]sets.String)
+	for p, cmap := range a.registry {
+		if activePods.Has(p) {
+			for _, rmap := range cmap {
+				for r, dinfo := range rmap {
+					if _, ok := ret[r]; !ok {
+						ret[r] = sets.NewString()
+					}
+					ret[r] = ret[r].Union(dinfo.ids)
+				}
+			}
+			continue
+		} else {
+			delete(a.registry, p)
+			glog.V(4).Infof("removed inactive pod: %v", p)
+		}
 	}
+	return ret
 }
 
-func (pdev podDevices) delete(pods []string) {
-	for _, uid := range pods {
-		delete(pdev, uid)
+// reserve processes the device requests from a Pod and return the set of
+// device IDs to allocate.
+func (a *devAllocationImpl) reserve(allDevices map[string]sets.String, req *allocRequest) (map[string]sets.String, error) {
+	a.Lock()
+	defer a.Unlock()
+
+	maxCount := make(map[string]int)
+	for container, rmap := range req.inits {
+		for resource, count := range rmap {
```

Comment:
We probably should either keep Init container re-use out of this PR or until we have settled onto a solution in PR #56818

Reply:
Well ... I have been following those discussions closely. There are many threads in that discussion: device identification, device (re)allocation, device reuse, device sharing, ... All important topics that needs a careful design and fruitful discussions. However, in this PR, I was only trying to do one thing: allocate devices to Pod rather than individual containers. When we decide in future that devices should have IDs or unique names, we can improve the framework so that a container can name a specific device to use.
Another thread orthogonal to this is about "preparation" before "use". When such an API is in place, we can improve container manager to invoke that call.

I'm okay with removing the re-use logic from this PR. However, we still need to make sure the allocation is done right. If the init-containers are requesting 2 devices and the regular containers are requesting 2 devices, we should have the Pod allocate at most 2 devices. In other words, the following should be enforced in the first place. The current implementation that adds up all device requests is not acceptable because it breaks k8s design.

```
device requests = max(max(req_by_init_containers), sum(req_by_regular_containers).
```

There are several usage patterns in my mind/opinion which are somehow equally important:

Scenario 1:

- devices are prepared and used by a regular container
- regular container developers have full knowledge about device interaction
- a separate init-container is useless and it may break the assumption

```
  containers:
   - name: compute1
      resources:
         vendor.com/accl: 2
```

Scenario 2:

- devices (transparently, blindly) prepared by init-containers, this may include preparing libraries, scripts, binaries for the regular containers. It may and may not involve device interaction. If these init-containers do touch the devices, a protocol must be in place what it can do.
- regular containers are designed to interact with devices
- neither init-containers nor regular containers care which devices they are using.
- here we assume all devices used are properly prepared by init-containers.

```
  init_containers:
    - name: c1
       resources:
          vendor.com/accl: 2
  containers:
   - name: compute1
      resources:
         vendor.com/accl: 1
   - name: compute2
      resources:
         vendor.com/accl: 1
```

Scenario 3:

- devices (transparently) prepared by (privileged) init-containers
- regular containers are NOT designed to directly interact with the devices, they will use the devices in other ways: a new UNIX domain socket, a new device volume for read/write, a SDK that abstracts away the function accelerated by the devices, a new network interface ... etc.
- neither init-containers nor regular containers care which devices they are using.

```
  init_containers:
    - name: c1
       resources:
          vendor.com/accl: 2
  containers:
   - name: compute1
      deviceMounts:
         - path: /dev/accel_1
         - path: /dev/accel_2
```

Scenario 4:

* This is a **sick** configuration, but ... it may still be valid *
- similar to scenario 1, but number of devices used by init-containers is **not** equal to the number of devices "consumed" by regular containers.
- currently we have no control over the mapping between "prepared" devices to "consumed" devices. Devices "A" and "B" prepared may be distributed to compute1 and computer2 in different ways.
- if wanted we can reject this pattern during validation

```
  init_containers:
    - name: c1
       resources:
          vendor.com/accl: 2
  containers:
   - name: compute1
      resources:
         vendor.com/accl: 2
   - name: compute2
      resources:
         vendor.com/accl: 1
```

Scenario 5:

- Devices are (transparently) prepared by different init-containers, for whatever good reason
- Regular containers consume them with and without realizing they have been prepared in different ways.
- This configuration should validate, IMO. 

```
  init_containers:
    - name: c1
       resources:
          vendor.com/accl: 1
    - name: c2
       resources:
          vendor.com/accl: 1
  containers:
   - name: compute1
      resources:
         vendor.com/accl: 2
```

Reply:
+1 on keeping device re-use changes separate from this PR.

Reply:
Thanks for the suggestion. So, to clarify, the suggestion is to keep the current behavior: calculate device requests as: 

```
sum(sum(request-from-init-containers) + sum(requests-from-regular-containers))
```
right?

Reply:
Sounds good :)

Reply:
wow ... device reuse logic is merged ...
Is there still a need to revert this to the old behavior now?